sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/parser/DebugExpr.atg

```diff
@@ -0,0 +1,396 @@
+import com.oracle.truffle.llvm.runtime.types.Type;
+import com.oracle.truffle.llvm.runtime.types.PrimitiveType;
+import com.oracle.truffle.llvm.runtime.types.VoidType;
+import org.graalvm.collections.Pair;
+// Set the name of your grammar here (and at the end of this grammar):
+COMPILER DebugExpr
+
+// Add auxiliary methods and declaration here.
+boolean IsCast() {
+	Token peek = scanner.Peek();
+	if(la.kind==_lpar) {
+	    while(peek.kind==_asterisc) peek=scanner.Peek();
+	    int k = peek.kind;
+	    if(k==_signed||k==_unsigned||k==_int||k==_long||k==_char||k==_short||k==_float||k==_double) return true;
+	}
+	return false;
+}
+
+private Iterable<Scope> scopes;
+private LLVMExpressionNode astRoot=null;
+private ContextReference<LLVMContext> contextReference=null;
+private DebugExprNodeFactory NF;
+public final static DebugExprErrorNode noObjNode = DebugExprErrorNode.create("<cannot find expression>");
+public final static DebugExprErrorNode errorObjNode = DebugExprErrorNode.create("<cannot evaluate expression>");
+
+void SetScopes(Iterable<Scope> scopes) {
+	this.scopes = scopes;
+}
+
+void SetContextReference(ContextReference<LLVMContext> contextReference) {
+	this.contextReference = contextReference;
+	NF = DebugExprNodeFactory.getInstance(contextReference);
+}
+
+public int GetErrors() {
+	return errors.count;
+}
+
+public LLVMExpressionNode GetASTRoot() {return astRoot; }
+// If you want your generated compiler case insensitive add the
+// keyword IGNORECASE here.
+
+
+CHARACTERS
+// Add character set declarations here.
+// Examples:
+   letter = 'A'..'Z' + 'a'..'z'.
+   digit = "0123456789".
+   cr = '\r'.
+   lf = '\n'.
+   singlecomma = '\''.
+   quote = '"'.
+
+
+TOKENS
+// Add token declarations here.
+// Example:
+   ident = letter {letter | digit}.
+   number = digit {digit}.
+   floatnumber = digit {digit} "." digit {digit} [ ("e"|"E") ["+"|"-"] digit {digit}].
+   charConst = singlecomma (letter|digit) singlecomma.
+   stringType = quote {letter|digit} quote.
+   lpar = '('.
+   asterisc = '*'.
+   signed = "signed".
+   unsigned = "unsigned".
+   int = "int".
+   long = "long".
+   short = "short".
+   float = "float".
+   double = "double".
+   char = "char".
+
+PRAGMAS
+// Add pragma declarations here.
+// Example:
+//   switch = '$' { digit | letter }. (. Optional semantic action .)
+
+
+// Add comments here.
+// Example for a multi-line block comment:
+COMMENTS FROM "/*" TO "*/" NESTED
+// Example for a single line comment:
+   COMMENTS FROM "//" TO lf
+
+
+// Set the ignored characters (whitespaces) here, the blank character is
+// ignored by default.
+// Example, add line breaks to the ignore set.
+IGNORE cr + lf
+
+
+PRODUCTIONS
+
+// Add your productions here, one must have the same name as the grammar,
+// it will be the start symbol (entry point of your compiler).
+// Example:
+//   Empty = "BEGIN" { Statement } "END".
+//   Statement = ident "=" number { "+" number } .
+DebugExpr		(. LLVMExpressionNode n=null; .)
```

Comment:
Please re-format this ATG so that all code attributes (the `(. ... .)`) begin at the same column.